server/src/main/java/org/elasticsearch/repositories/GetSnapshotInfoContext.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+package org.elasticsearch.repositories;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.util.concurrent.CountDown;
+import org.elasticsearch.snapshots.SnapshotId;
+import org.elasticsearch.snapshots.SnapshotInfo;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.function.BiConsumer;
+import java.util.function.BooleanSupplier;
+
+/**
+ * Describes the context of fetching one or more {@link SnapshotInfo} via {@link Repository#getSnapshotInfo(GetSnapshotInfoContext)}.
+ */
+public final class GetSnapshotInfoContext implements ActionListener<SnapshotInfo> {
+
+    private static final Logger logger = LogManager.getLogger(GetSnapshotInfoContext.class);
+
+    /**
+     * Snapshot ids to fetch info for.
+     */
+    private final List<SnapshotId> snapshotIds;
+
+    /**
+     * Stop fetching additional {@link SnapshotInfo} if an exception is encountered.
+     */
+    private final boolean abortOnFailure;
+
+    /**
+     * If this supplier returns true, indicates that the task that initiated this context has been cancelled and that not further fetching
+     * of {@link SnapshotInfo} should be started.
+     */
+    private final BooleanSupplier isCancelled;
+
+    /**
+     * Listener resolved when fetching {@link SnapshotInfo} has completed. If resolved successfully, no more calls to
+     * {@link #onSnapshotInfo} will be made. Only resolves exceptionally if {@link #abortOnFailure} is true in case one or more
+     * {@link SnapshotInfo} failed to be fetched.
+     */
+    private final ActionListener<Void> doneListener;
+
+    /**
+     * {@link BiConsumer} invoked for each {@link SnapshotInfo} that is fetched with this instance and the {@code SnapshotInfo} as
+     * arguments.
+     */
+    private final BiConsumer<GetSnapshotInfoContext, SnapshotInfo> onSnapshotInfo;
+
+    private final CountDown counter;
+
+    public GetSnapshotInfoContext(
+        Collection<SnapshotId> snapshotIds,
+        boolean abortOnFailure,
+        BooleanSupplier isCancelled,
+        BiConsumer<GetSnapshotInfoContext, SnapshotInfo> onSnapshotInfo,
+        ActionListener<Void> listener
+    ) {
+        assert snapshotIds.isEmpty() == false : "no snapshot ids to fetch given";
```

Comment:
IMO we should either forbid this in production too (e.g. throw an `IllegalArgumentException`) or else just allow it and deal with the `n=0` case ourselves. I'd rather the latter I think but either way is ok.

Reply:
I went with throwing for now. I think we shouldn't create the case where we pass an empty list to this API, just like we don't allow a 0 sized `GroupedActionListener`, otherwise we'd have to resolve the listener in the constructor right away I guess, which I find confusing.